### PR TITLE
Tooling updates for dealing with native services distributed in pub packages

### PR DESCRIPTION
**Pull request by chinmaygarde**
cc @devoncarew @mpcomplete This now allows us to load a pub package like https://pub.dartlang.org/packages/firebase_flutter

### CODE REVIEW

**Comment by chinmaygarde** [comment]
This requires a small patch in the engine https://github.com/flutter/engine/pull/2394

**Comment by devoncarew** [comment]
I'll take a quick spin through but @mpcomplete will be more familiar w/ what's going on.

File: packages/flutter_tools/lib/src/services.dart [hunk 1]
```diff
@@ -10,10 +10,13 @@ import 'package:path/path.dart' as path;
 import 'package:yaml/yaml.dart';
 
 import 'artifacts.dart';
+import 'base/globals.dart';
 
 const String _kFlutterManifestPath = 'flutter.yaml';
+const String _kFlutterServicesManifestPath = 'flutter_services.yaml';
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/services.dart]: So this is a metadata file in each package that contains flutter services?

File: packages/flutter_tools/lib/src/services.dart [hunk 2]
```diff
@@ -77,22 +86,16 @@ Future<String> getServiceFromUrl(
 ///   ]
 /// }
 File generateServiceDefinitions(
-  String dir, List<Map<String, String>> servicesIn, { bool ios }
+  String dir, List<Map<String, String>> servicesIn
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/services.dart]: nit: this can be one line now

**Comment by chinmaygarde** [comment]
Beginnings of some documentation for flutter_services.yaml  at https://docs.google.com/document/d/1SPhelwt0wMgUwHQXi6-2lzvqOtXxE4Y0J3DB33JL5ac/edit#heading=h.ahz5yayw3njg

File: packages/flutter_tools/lib/src/ios/device_ios.dart [hunk 3]
```diff
@@ -546,52 +546,48 @@ Future<bool> _buildIOSXcodeProject(ApplicationPackage app, bool isDevice) async
...
-
-    generateServiceDefinitions(bundle.path, services, ios: true);
-  }
+  List<Map<String, String>> services = [];
+  printTrace("Trying to resolve native pub services");
+
+  // Step 1: Parse the service configuration yaml files present in the service
+  //         pub packages.
+  await parseServiceConfigs(services);
+  printTrace("Found ${services.length} service definition(s)");
```
Comment (mpcomplete) [inline packages/flutter_tools/lib/src/ios/device_ios.dart]: Put .'s at the end of all the print statements.

File: packages/flutter_tools/lib/src/ios/device_ios.dart [hunk 4]
```diff
@@ -546,52 +546,48 @@ Future<bool> _buildIOSXcodeProject(ApplicationPackage app, bool isDevice) async
...
+void _copyServiceDefinitionsManifest(List<Map<String, String>> services, File manifest) {
+  printTrace("Creating service definitions manifest at '${manifest.path}'");
+  List<Map<String, String>> jsonServices = services.map((Map<String, String> service) => {
+    'name': service['name'],
+    // Since we have already moved it to the Frameworks directory. Strip away
+    // the directory and basenames.
+    'framework': path.basenameWithoutExtension(service['ios-framework'])
+  }).toList();
+  Map<String, dynamic> json = { 'services' : jsonServices };
+  manifest.writeAsStringSync(JSON.encode(json), mode: FileMode.WRITE, flush: true);
```
Comment (mpcomplete) [inline packages/flutter_tools/lib/src/ios/device_ios.dart]: I'd prefer this to be moved back to services.json to be shared with the android code. We should also use the same filename on both platforms.

File: packages/flutter_tools/lib/src/ios/device_ios.dart [hunk 5]
```diff
@@ -546,52 +546,48 @@ Future<bool> _buildIOSXcodeProject(ApplicationPackage app, bool isDevice) async
...
-  }
-}
-
-void _copyFrameworksToBundle(String destDir, List<Map<String, String>> services) {
-  // TODO(mpcomplete): check timestamps.
-  for (Map<String, String> service in services) {
-    String basename = path.basename(service['framework-path']);
-    String destPath = path.join(destDir, basename);
-    _copyDirRecursive(service['framework-path'], destPath);
+    File dylib = new File(path.join(service['root'], service['ios-framework']));
```
Comment (mpcomplete) [inline packages/flutter_tools/lib/src/ios/device_ios.dart]: Don't do this. Use the getServiceFromUrl helper from services.json. It knows how to fetch a remote URL and cache it locally. (It returns the path to the local file).

**Comment by chinmaygarde** [comment]
Will move the ServiceDefinitions.json updates into another patch. Other comments have been addressed. Another look please.

File: packages/flutter_tools/lib/src/ios/device_ios.dart [hunk 6]
```diff
@@ -546,52 +546,49 @@ Future<bool> _buildIOSXcodeProject(ApplicationPackage app, bool isDevice) async
...
-    _copyDirRecursive(service['framework-path'], destPath);
+    String dylibPath = await getServiceFromUrl(service['ios-framework'], service['root'], service['name']);
+    File dylib = new File(dylibPath);
+    printTrace("Copying ${dylib.path} into bundle.");
+    if (!dylib.existsSync()) {
+      printError("The service dylib '${dylib.path}' does not exist.");
+      continue;
+    }
+    // Shell out so permissions on the dylib are preserved.
+    runCheckedSync(['/bin/cp', dylib.path, frameworksDirectory.path]);
```
Comment (mpcomplete) [inline packages/flutter_tools/lib/src/ios/device_ios.dart]: Does 'cp' on mac preserve permissions automatically? You need a '-a' arg to do that on linux.

**Comment by mpcomplete** [comment]
LGTM.

I'm not sure why it's easier to split the ios/android services.json generation this patch, then merge them again in a separate patch later, but whatever :).

**Pull request merged**